src/__experimental__/components/decimal/decimal.component.js

```diff
@@ -0,0 +1,123 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import Textbox from '../textbox';
+import I18nHelper from '../../../utils/helpers/i18n';
+import Logger from '../../../utils/logger';
+
+class Decimal extends React.Component {
+  input = null;
+
+  // Create ref to document for tests
+  _document = document;
+
+  formatValue = () => {
+    const { value } = this.props;
+    const { input } = this;
+
+    // Return unformatted value if component has not mounted
+    if (!input) {
+      return value;
+    }
+    // Return unformatted value if input is being used
+    if (input && this._document.activeElement === input.current) {
+      return value;
+    }
+
+    // Only format value if input is not active
+    // Strip delimiters otherwise formatDecimal Helper goes nuts
+    const format = I18nHelper.format();
+    const delimiter = `\\${format.delimiter}`;
+    const removeDelimiter = new RegExp(`[${delimiter}]*`, 'g');
+    const noDelimiters = value.replace(removeDelimiter, '');
+
+    return I18nHelper.formatDecimal(
+      noDelimiters,
+      this.validatePrecision()
+    );
+  }
+
+  validatePrecision = () => {
+    const { precision } = this.props;
+
+    if (precision > 15) {
+      Logger.warn('Precision cannot be greater than 15');
+      return 15;
+    }
+
+    return precision;
+  }
+
+  testValue = (value) => {
+    const { precision } = this.props;
+    const format = I18nHelper.format();
+    const delimiter = `\\${format.delimiter}`;
+    const seperator = `\\${format.separator}`;
+    const isGoodDecimal = new RegExp(`^[\\d${delimiter}]*[${seperator}{1}]?\\d{0,${precision}}?$`);
+
+    return isGoodDecimal.test(value);
+  }
+
+  onChange = (evt) => {
+    const { target } = evt;
+    const { value, selectionEnd } = evt.target;
+    const isValid = this.testValue(value);
+
+    if (isValid) {
+      this.props.onChange(evt);
+
+      setTimeout(() => {
+        target.setSelectionRange(selectionEnd, selectionEnd);
```

Comment:
I'm not sure this is the correct place for this to happen. Surely the cursor is in the correct position while active and valid.

Looking at the old implementation it only sets the cursor if the value is invalid, in which it replaces the value with the previously valid value https://github.com/Sage/carbon/blob/master/src/components/decimal/decimal.js#L153-L161

Currently in your implementation if you try it in the browser and type really fast it breaks the cursor position.

Reply:
Think it should be
```
    if (isValid) {
      this.props.onChange(evt);
    } else {
      target.setSelectionRange(selectionEnd, selectionEnd);
    }
```